include/oneapi/dpl/pstl/algorithm_ranges_impl.h

```diff
@@ -349,6 +350,325 @@ __pattern_search_n(_Tag __tag, _ExecutionPolicy&& __exec, _R&& __r, std::ranges:
+          typename _Proj>
+auto
+__pattern_copy_if_impl(_Tag __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r, _Pred __pred, _Proj __proj)
+{
+    static_assert(__is_parallel_tag_v<_Tag> || typename _Tag::__is_vector{});
+
+    auto __pred_1 = [__pred, __proj](auto&& __val) { return std::invoke(__pred, std::invoke(__proj,
+        std::forward<decltype(__val)>(__val)));};
+
+    auto __res_idx = oneapi::dpl::__internal::__pattern_copy_if(__tag, std::forward<_ExecutionPolicy>(__exec),
+        std::ranges::begin(__in_r), std::ranges::begin(__in_r) + std::ranges::size(__in_r),
+        std::ranges::begin(__out_r), __pred_1) - std::ranges::begin(__out_r);
+
+    using __return_t = std::ranges::copy_if_result<std::ranges::borrowed_iterator_t<_InRange>,
+        std::ranges::borrowed_iterator_t<_OutRange>>;
+
+    return __return_t{std::ranges::begin(__in_r) + std::ranges::size(__in_r), __out_r.begin() + __res_idx};
+}
+
+template <typename _IsVector, typename _ExecutionPolicy, typename _InRange, typename _OutRange, typename _Pred,
+          typename _Proj>
+auto
+__pattern_copy_if_ranges(__parallel_tag<_IsVector> __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r,
+                         _Pred __pred, _Proj __proj)
+{
+    return __pattern_copy_if_impl(__tag, std::forward<_ExecutionPolicy>(__exec), std::forward<_InRange>(__in_r),
+        std::forward<_OutRange>(__out_r), __pred, __proj);
+}
+
+template<typename _Tag, typename _ExecutionPolicy, typename _InRange, typename _OutRange, typename _Pred, typename _Proj,
+             oneapi::dpl::__internal::__enable_if_execution_policy<_ExecutionPolicy, int> = 0>
+auto
+__pattern_copy_if_ranges(_Tag __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r, _Pred __pred, _Proj __proj)
+{
+    if constexpr(typename _Tag::__is_vector{})
+        return __pattern_copy_if_impl(__tag, std::forward<_ExecutionPolicy>(__exec), std::forward<_InRange>(__in_r),
+            std::forward<_OutRange>(__out_r), __pred, __proj);
+    else
+        return std::ranges::copy_if(std::forward<_InRange>(__in_r), std::ranges::begin(__out_r), __pred, __proj);
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_merge
+//---------------------------------------------------------------------------------------------------------------------
+
+template<typename _Tag, typename _ExecutionPolicy, typename _R1, typename _R2, typename _OutRange, typename _Comp,
+         typename _Proj1, typename _Proj2>
+auto
+__pattern_merge_impl(_Tag __tag, _ExecutionPolicy&& __exec, _R1&& __r1, _R2&& __r2, _OutRange&& __out_r, _Comp __comp,
+                     _Proj1 __proj1, _Proj2 __proj2)
+{
+    static_assert(__is_parallel_tag_v<_Tag> || typename _Tag::__is_vector{});
+    assert(std::ranges::size(__r1) + std::ranges::size(__r2) == std::ranges::size(__out_r));
```

Comment:
Similar question here about range sizes. Do we need `std::ranges::size(__r1) + std::ranges::size(__r2) == std::ranges::size(__out_r)` or `std::ranges::size(__r1) + std::ranges::size(__r2) <= std::ranges::size(__out_r)`?

Reply:
Yes, it looks reasonable to make the condition weaker.